### PR TITLE
Add new aggregate function sumMap(key, value).

### PR DESCRIPTION
**Pull request by bocharov**
Adding new aggregate function **sumMap**, which takes two arguments: keys and values, and as a result, builds an array of 2 arrays - ordered keys and values summed up  by corresponding keys.

This function is the most useful when using SummingMergeTree to sum Nested columns, which name ends in "Map".

Example: sumMap(k, v) of:
```
 k           v
 [1,2,3]     [10,10,10]
 [3,4,5]     [10,10,10]
 [4,5,6]     [10,10,10]
 [6,7,8]     [10,10,10]
 [7,5,3]     [5,15,25]
 [8,9,10]    [20,20,20]
```

will return:
```
([1,2,3,4,5,6,7,8,9,10],[10,10,45,20,35,20,15,30,20,20])
```

### CODE REVIEW

File: dbms/src/AggregateFunctions/AggregateFunctionSumMap.h
Comment (alexey-milovidov) [inline dbms/src/AggregateFunctions/AggregateFunctionSumMap.h]: This is not suitable, because:
- arbitary key types are not supported;
- keys will be serialized with garbage.

Better to just write a loop and to use `IDataType::serializeBinary` method.

File: dbms/src/AggregateFunctions/AggregateFunctionSumMap.h
Comment (alexey-milovidov) [inline dbms/src/AggregateFunctions/AggregateFunctionSumMap.h]: Excessive copy. Better to insert into nested column directly.

File: dbms/src/Core/Field.h
Comment (alexey-milovidov) [inline dbms/src/Core/Field.h]: Better to use `FieldVisitorSum` directly inside `AggregateFunctionSumMap`. You can move it out of `SummingSortedBlockInputStream.cpp` to `FieldVisitors.h` or a separate header file.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
Almost fine, but few minor changes required.

**Comment by alexey-milovidov** [comment]
Implementation is rather inefficient, also we lack limit on array sizes and memory tracking. This is Ok for first version (will merge after few modifications). Superior implementation can use implementation similar to `AggregateFunctionGroupUniqArray`.

File: dbms/src/AggregateFunctions/AggregateFunctionSumMap.h
Comment (bocharov) [reply dbms/src/AggregateFunctions/AggregateFunctionSumMap.h]: It makes sense to serialize as little as possible. The only question I have - how to get access to data type of keys and values? In method `setArgumentsImpl` I can get access only to data type of arguments itself and that is `DataTypeArray`, but I need to know exact data type of the array elements. Is there a way to do this?

Comment (alexey-milovidov) [reply dbms/src/AggregateFunctions/AggregateFunctionSumMap.h]: You can obtain nested type from DataTypeArray: `DataTypeArray::getNestedType`.

Comment (bocharov) [reply dbms/src/AggregateFunctions/AggregateFunctionSumMap.h]: Thanks! It worked, new serialization method consumes much less memory now.

File: dbms/src/Core/Field.h
Comment (bocharov) [reply dbms/src/Core/Field.h]: Fixed, moved `FieldVisitorSum` to `FieldVisitors.h`, though not sure if that more efficient than using overloaded operator `+=`.

File: dbms/src/AggregateFunctions/AggregateFunctionSumMap.h
Comment (bocharov) [reply dbms/src/AggregateFunctions/AggregateFunctionSumMap.h]: I've tried to do that but couldn't access sub-arrays of `ColumnArray`. I also tried to find similar implementation with usage of `getNestedColumn` but existing examples are quite different. There is no `NestedColumn` column type.

Can you please explain bit more how to implement the suggested change?

Comment (alexey-milovidov) [reply dbms/src/AggregateFunctions/AggregateFunctionSumMap.h]: Return type is `Array(Array(T))`

`auto & to_data = to_array.getData();`

`to_data` is elements of result arrays, so `Array(T)`.

Now if you get `static_cast<DataTypeArray &>(to_data).getData()`, you will get column of type `T` that will contain all elements of all arrays.

And you can insert individual elements into it, one by one.

Comment (alexey-milovidov) [reply dbms/src/AggregateFunctions/AggregateFunctionSumMap.h]: PS. Why return type is not `Tuple(Array(K), Array(V))`?
This will allow different types of keys and values.
For example, keys can be strings and values - integers.
And manipulation with the result will be more simple.

Better to change signature before merge.

**Comment by bocharov** [comment]
I changed return type to `Tuple(Array(K), Array(V))` and did other requested changes.
Please review again and I hope we could merge then.

File: dbms/src/AggregateFunctions/AggregateFunctionSumMap.h
Comment (alexey-milovidov) [inline dbms/src/AggregateFunctions/AggregateFunctionSumMap.h]: Why we reserve size for just inserted array even if column can already have multiple array values?

Comment (bocharov) [reply dbms/src/AggregateFunctions/AggregateFunctionSumMap.h]: What do you mean? I do not see just inserted array. `to_keys` and `to_values` are of the type ColumnVector<T> and I thought allocating memory for vector elements before inserting them will be more efficient. Please feel free to remove this or change it to something more efficient.

Comment (alexey-milovidov) [reply dbms/src/AggregateFunctions/AggregateFunctionSumMap.h]: `to_keys_col` and `to_values_col` could already contain many arrays, but you reserve just for one array.

**Pull request merged**
